rfcs/20180920-unify-rnn-interface.md

```diff
@@ -0,0 +1,377 @@
+|BidirectionalGridLSTMCell|low usage within Google|Leave in v1.compat due to low usage|
+|NASCell|low usage within Google|Leave in v1.compat due to low usage|
+|UGRNNCell|low usage within Google|Leave in v1.compat due to low usage|
+|IntersectionRNNCell|low usage within Google|Leave in v1.compat due to low usage|
+|PhasedLSTMCell|low usage within Google|Leave in v1.compat due to low usage|
+|ConvLSTMCell<br>Conv1DLSTMCell<br>Conv2DLSTMCell<br>Conv3DLSTMCell|low usage within Google|Keras has a Conv2DLSTM implementation and this is more generic. Probably port this core|
+|GLSTMCell|low usage within Google|Leave in v1.compat due to low usage|
+|SRUCell|low usage within Google|Leave in v1.compat due to low usage|
+|IndRNNCell|low usage within Google|Leave in v1.compat due to low usage|
+|IndyGRUCell|low usage within Google|Leave in v1.compat due to low usage|
+|IndyLSTMCell|low usage within Google|Leave in v1.compat due to low usage|
+|AttentionCellWrapper|Medium usage|Move to core|
+|HighwayWrapper|low usage within Google|Leave in v1.compat due to low usage|
+
+
+### CuDNN implementation vs Normal implementation
+
+NVidia published the CuDNN support for RNN in the [blog post](https://devblogs.nvidia.com/optimizing-recurrent-neural-networks-cudnn-5/), 
+which provides a fast underlying implementation for LSTM and GRU. The downside however, is that it 
+does not support all the features that extended from various varieties of LSTM and GRU.
+
+Currently both TF RNN and Keras RNN are exposing CuDNN cells as different type from the basic LSTM 
+and GRU cell. One of the goals here is to hide this details from user and choose smartly for user. 
+This idea is original proposed by drpng@.
+
+The CuDNN implementation fused several inputs and internal steps together to achieve the performance
+gain:
+
+1. Merge 4 internal matmul into 1 within LSTM cell
+1. the timestep loop
+1. multiple layers stacked together
+1. bidirectional RNN
+
+It also has few differences from the original LSTM/GRU implementation:
+
+1. The output projection in CuDNN has an extra bias, which cause the weights of the CuDNN 
+incompatible with the standard LSTM/GRU. There are internal effort to convert the weights between 
+a CuDNN implementation and normal TF implementation. See CudnnLSTMSaveable.
+1. CuDNN does not support variational recurrent dropout, which is a quite important feature.
+1. CuDNN implementation only support TAN activation which is also the default implementation in the 
+LSTM paper. The Keras one support more activation choices if user don't want the default behavior.
+
+With that, it means when users specify their LSTM/GRU layer, the underlying implementation could be 
+swapped to use CuDNN version if it meets certain criteria:
+
+```python
+class LSTM(RNN):
+  def __init__(self, .....):
+     if activation == 'tan' and dropout == 0 and use_bias == True:
+       self.could_use_cudnn = True
+     else:
+       self.could_use_cudnn = False
+     ....
```

Comment:
Will this impact the default behavior of the Cells? And in particular, the GRUCell (cuDNN support only 1406.1078v1 when Keras default is 1406.1078v3 and TF only supports 1406.1078v3) ?

Reply:
I don't think it will change the behavior. I am not so sure about the difference between 1406.1078v1 and v3, but I think one of them are the dropout defaults. I think you can also translate the code into:

if parameter_match_v1:
  self.could_use_cudnn = True
else
  self.could_use_cudnn = False

Reply:
ok thanks for the clarification